### ISSUE TITLE
Race condition when a CSI volume with staging gets unmounted from one alloc and mounted to another alloc

### ISSUE THREAD

**Issue by ygg-drop** [hashicorp/nomad#20424]
### Nomad version
Output from `nomad version`

```
Nomad v1.7.6+ent
BuildDate 2024-03-12T07:55:51Z
Revision a71f7e82ef60200f273a92601b1b366429b0bd5f
```

### Operating system and Environment details

Alpine 3.19.1

### Issue

Let's assume there are two allocations - A and B - and both of them use the same CSI volume with staging and `multi-node-multi-writer` access mode (CephFS in this case).

The following events happen:

1. alloc A has the volume mounted
2. alloc A is stopped and Nomad calls `NodeUnpublishVolume` RPC to the CSI driver followed by `NodeUnstageVolume`
3. at the same time alloc B is started and Nomad calls `NodePublishVolume` on the same volume, **missing the fact that the same volume was unstaged with alloc A**.
4. in the CSI driver, the `NodePublishVolume` RPC finishes first and the volume gets bind-mounted from `staging` path to `per-alloc` path
5. in the CSI driver, the `NodeUnstageVolume` RPC finishes second, and the volume gets unmounted from `staging` path, which results in alloc B's `per-alloc` bind mount **now pointing to the underlying local device instead of the network volume**.
6. subsequent calls to  `NodeUnpublishVolume` RPC for alloc B fail with `device or resource busy` error.

Below is a sanitized `cephcsi` node plugin's log fragment capturing just the relevant events. In this case both allocations were failing due to an unrelated error which resulted in multiple rescheduling of the jobs, which probably increased the likelihood of this bug to occur.

- `VOLUME_A` - the volume used by both allocs
- `b2bd2806-f675-41f7-8efc-8a27830ce3d1` - alloc A
- `5a77c319-f7ab-4fd5-98cb-1b73538f735f` - alloc B

```
I0101 01:04:27.675127       1 utils.go:195] ID: 11620 Req-ID: VOLUME_A GRPC call: /csi.v1.Node/NodeUnpublishVolume
I0101 01:04:27.675173       1 utils.go:206] ID: 11620 Req-ID: VOLUME_A GRPC request: {"target_path":"/local/csi/per-alloc/b2bd2806-f675-41f7-8efc-8a27830ce3d1/VOLUME_A/rw-file-system-multi-node-multi-writer","volume_id":"VOLUME_A"}
I0101 01:04:27.677395       1 cephcmds.go:105] ID: 11620 Req-ID: VOLUME_A command succeeded: umount [/local/csi/per-alloc/b2bd2806-f675-41f7-8efc-8a27830ce3d1/VOLUME_A/rw-file-system-multi-node-multi-writer]
I0101 01:04:27.677462       1 nodeserver.go:586] ID: 11620 Req-ID: VOLUME_A cephfs: successfully unbounded volume VOLUME_A from /local/csi/per-alloc/b2bd2806-f675-41f7-8efc-8a27830ce3d1/VOLUME_A/rw-file-system-multi-node-multi-writer
I0101 01:04:27.677485       1 utils.go:212] ID: 11620 Req-ID: VOLUME_A GRPC response: {}
I0101 01:04:27.677753       1 utils.go:195] ID: 11621 Req-ID: VOLUME_A GRPC call: /csi.v1.Node/NodeUnstageVolume
I0101 01:04:27.677783       1 utils.go:206] ID: 11621 Req-ID: VOLUME_A GRPC request: {"staging_target_path":"/local/csi/staging/VOLUME_A/rw-file-system-multi-node-multi-writer","volume_id":"VOLUME_A"}
I0101 01:04:27.679307       1 utils.go:195] ID: 11623 Req-ID: VOLUME_A GRPC call: /csi.v1.Node/NodePublishVolume
I0101 01:04:27.679398       1 utils.go:206] ID: 11623 Req-ID: VOLUME_A GRPC request: {"secrets":"***stripped***","staging_target_path":"/local/csi/staging/VOLUME_A/rw-file-system-multi-node-multi-writer","target_path":"/local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer"
I0101 01:04:27.681349       1 cephcmds.go:105] ID: 11623 Req-ID: VOLUME_A command succeeded: mount [-o bind,_netdev /local/csi/staging/VOLUME_A/rw-file-system-multi-node-multi-writer /local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer]
I0101 01:04:27.681364       1 nodeserver.go:530] ID: 11623 Req-ID: VOLUME_A cephfs: successfully bind-mounted volume VOLUME_A to /local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer
I0101 01:04:27.681378       1 utils.go:212] ID: 11623 Req-ID: VOLUME_A GRPC response: {}
I0101 01:04:27.681868       1 volumemounter.go:126] requested mounter: , chosen mounter: kernel
I0101 01:04:27.681915       1 cephcmds.go:105] ID: 11621 Req-ID: VOLUME_A command succeeded: umount [/local/csi/staging/VOLUME_A/rw-file-system-multi-node-multi-writer --all-targets]
I0101 01:04:27.681930       1 nodeserver.go:647] ID: 11621 Req-ID: VOLUME_A cephfs: successfully unmounted volume VOLUME_A from /local/csi/staging/VOLUME_A/rw-file-system-multi-node-multi-writer
I0101 01:04:33.997444       1 utils.go:195] ID: 11627 Req-ID: VOLUME_A GRPC call: /csi.v1.Node/NodeUnpublishVolume
I0101 01:04:33.997495       1 utils.go:206] ID: 11627 Req-ID: VOLUME_A GRPC request: {"target_path":"/local/csi/per-alloc/b2bd2806-f675-41f7-8efc-8a27830ce3d1/VOLUME_A/rw-file-system-multi-node-multi-writer","volume_id":"VOLUME_A"}
E0101 01:04:33.997524       1 nodeserver.go:550] ID: 11627 Req-ID: VOLUME_A stat failed: stat /local/csi/per-alloc/b2bd2806-f675-41f7-8efc-8a27830ce3d1/VOLUME_A/rw-file-system-multi-node-multi-writer: no such file or directory
I0101 01:04:33.997534       1 nodeserver.go:554] ID: 11627 Req-ID: VOLUME_A targetPath: /local/csi/per-alloc/b2bd2806-f675-41f7-8efc-8a27830ce3d1/VOLUME_A/rw-file-system-multi-node-multi-writer has already been deleted
I0101 01:04:33.997546       1 utils.go:212] ID: 11627 Req-ID: VOLUME_A GRPC response: {}
I0101 01:05:35.405412       1 utils.go:195] ID: 11636 Req-ID: VOLUME_A GRPC call: /csi.v1.Node/NodeUnpublishVolume
I0101 01:05:35.405473       1 utils.go:206] ID: 11636 Req-ID: VOLUME_A GRPC request: {"target_path":"/local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer","volume_id":"VOLUME_A"}
E0101 01:05:35.405702       1 utils.go:210] ID: 11636 Req-ID: VOLUME_A GRPC error: rpc error: code = Internal desc = unlinkat /local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer: device or resource busy
I0101 01:05:35.406140       1 utils.go:195] ID: 11638 Req-ID: VOLUME_A GRPC call: /csi.v1.Node/NodeUnpublishVolume
I0101 01:05:35.406168       1 utils.go:206] ID: 11638 Req-ID: VOLUME_A GRPC request: {"target_path":"/local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer","volume_id":"VOLUME_A"}
E0101 01:05:35.406242       1 utils.go:210] ID: 11638 Req-ID: VOLUME_A GRPC error: rpc error: code = Internal desc = unlinkat /local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer: device or resource busy
I0101 01:05:37.407576       1 utils.go:195] ID: 11640 Req-ID: VOLUME_A GRPC call: /csi.v1.Node/NodeUnpublishVolume
I0101 01:05:37.407617       1 utils.go:206] ID: 11640 Req-ID: VOLUME_A GRPC request: {"target_path":"/local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer","volume_id":"VOLUME_A"}
E0101 01:05:37.407721       1 utils.go:210] ID: 11640 Req-ID: VOLUME_A GRPC error: rpc error: code = Internal desc = unlinkat /local/csi/per-alloc/5a77c319-f7ab-4fd5-98cb-1b73538f735f/VOLUME_A/rw-file-system-multi-node-multi-writer: device or resource busy
```

### Reproduction steps

Run two jobs which both use the same volume with `multi-node-multi-writer` access mode (e.g. CephFS). Configure the jobs to fail and get rescheduled on the same node with a short period which should increase the chance of this bug happening.

#### Expected Result

Nomad should ensure that a volume that got unstaged should not be published before getting staged again.

#### Actual Result

Nomad does not seem to ensure unstaged volumes can get published.

**Comment by tgross** [comment]
Hi @ygg-drop! Yeah, we have enforcement of single in-flight operation per volume at the server but looks like we're missing it in the client. 

At a _very_ quick look I suspect we could fix this by adding a mutex in the [volumeManager](https://github.com/hashicorp/nomad/blob/v1.7.7/client/pluginmanager/csimanager/volume.go#L64). But the `volumeManager` supports a single plugin and not a single volume, so simply locking all operations would mean we'd only be able to mount a single volume at a time for the whole plugin, which would let an allocation slow down the placement of other allocations. We'd definitely like to avoid that, so we might need a more fine-grained locking structure there.

I'll mark this as accepted and we'll try to get this done when we look at https://github.com/hashicorp/nomad/issues/18741, which I know is on our board.

**Comment by tgross** [comment]
I've had a look at this and the existing "usage tracker" is supposed to prevent exactly this problem. But we call `v.usageTracker.Claim` (ref [`volume.go#L264`](https://github.com/hashicorp/nomad/blob/v1.8.0-beta.1/client/pluginmanager/csimanager/volume.go#L264)) only after we've completed both staging and publishing, which isn't right. 

We need to move that before staging/publishing. A patch to do that is pretty easy. So I just need to write some tests to exercise it and make sure that we're freeing the usage tracker at the right moment if we get an error for staging/publishing. I'm in-progress on that.

**Comment by tgross** [comment]
Draft PR is up https://github.com/hashicorp/nomad/pull/20550

**Comment by tgross** [comment]
#20550 has been merged and will ship in Nomad 1.8.0 (with backports to supported versions)
